design/dedicated-osd-pod.md

```diff
@@ -0,0 +1,54 @@
+# Run OSD in its own Pod
+
+## TL;DR
+
+A one-OSD-per-Pod placement should be implemented to improve reliability and resource efficiency for Ceph OSD daemon.
+
+## Background
+
+Currently in Rook 0.7, Rook Operator starts a ReplicaSet to run [`rook osd`](https://github.com/rook/rook/blob/master/cmd/rook/osd.go) command (hereafter referred to as `OSD Provisioner`)  on each storage node. The ReplicaSet has just one replica. `OSD Provisioner` scans and prepares devices, creates OSD IDs and data directories or devices, generates Ceph configuration. At last, `OSD Provisioner` starts all Ceph OSD, i.e. `ceph-osd`, daemons in foreground and tracks `ceph-osd` processes.
+
+As observed, all Ceph OSDs are running in the same Pod.
+
+## Limitations
+
+The limitations of current design are:
+
+- Reliability issue. One Pod for all OSDs doesn't have the highest reliability nor efficiency. If the Pod is deleted, accidentally or during maintenence, all OSDs are down till the ReplicaSet restart. 
+- Efficiency issue. Resource limits cannot be set effectively on the OSDs since the number of osds per in the pod could vary from node to node. The operator cannot make decisions about the topology because it doesn't know in advance what devices are available on the nodes.
+- Process management issue. Rook's process management is very simple. Using Kubernetes pod management is much more reliable.
+
+
+A more comprehensive discussions can be found at [this issue](https://github.com/rook/rook/issues/1341).
+
+## Proposal   
+
+We propose the following OSD startup change.
+
+### Create new OSDs
+| Sequence |Rook Operator  | OSD Provisioner   | Ceph OSD Deployment | 
+|---|---|---|---|
+| 1  | Read devices from cluster CRD and create a Pod for OSD Provisioner for each storage node  |   |   |
+| 2  | Watch Configmap  | Prepare OSDs  |   |
+| 3  | Watch Configmap  |  Persist OSD ID, datapath, and node info in a per node Configmap | |
+| 4  | Detect Configmap change, parse Configmap, extract OSD info, construct OSD Pod command and arg | |
+| 5  | Create one deployment per OSD | |
+| 6  | | | Start `ceph-osd` Daemon one Pod per device |
+
+This change addresses the above limitations in the follow ways:
+- High reliability. Each `ceph-osd` daemon runs its own Pod, their restart and upgrade are by Kubernetes controllers. Upgrading OSD Provisioner Pod no longer restarts `ceph-osd` daemons.
+- More efficient resource requests. Once OSD Provisioner discovers and prepares all OSD devices, Rook Operator is informed of the topology and assigns appropriate resources to each Ceph OSD deployment.
```

Comment:
The OSD Provisioner cannot perform discovery and preparation in the same step. The discovery needs to happen before the operator instructs the devices to be prepared. The operator needs to know about the actual devices available before it can apply the desired devices from the CRD. This is where the Rook agent was suggested to do the discovery. And if the devices have already been made available with local storage, the agent wouldn't need to discover those.